godot-macros/src/docs.rs

```diff
@@ -39,52 +39,72 @@ pub fn make_inherent_impl_docs(
     fn pieces(
         functions: &[FuncDefinition],
         signals: &[SignalDefinition],
         constants: &[ConstDefinition],
-    ) -> Option<TokenStream> {
+    ) -> TokenStream {
+        let to_tagged = |s: String, tag: &str| -> String {
+            if s.is_empty() {
+                s
+            } else {
+                format!("<{tag}>{s}</{tag}>")
+            }
+        };
+
+        let signals_block = to_tagged(
+            signals
+                .iter()
+                .filter_map(make_signal_docs)
+                .collect::<String>(),
+            "signals",
+        );
+        let constants_block = to_tagged(
+            constants
+                .iter()
+                .map(|ConstDefinition { raw_constant }| raw_constant)
+                .filter_map(make_constant_docs)
+                .collect::<String>(),
+            "constants",
+        );
+
         let methods = functions
             .iter()
-            .map(make_method_docs)
-            .collect::<Option<String>>()?;
-        let signals = signals
-            .iter()
-            .map(make_signal_docs)
-            .collect::<Option<String>>()?;
-        let constants = constants
-            .iter()
-            .map(|ConstDefinition { raw_constant: x }| x)
-            .map(make_constant_docs)
-            .collect::<Option<String>>()?;
-        let field_definition = quote! {
+            .filter_map(make_method_docs)
+            .collect::<String>();
+
+        let methods = if methods.is_empty() {
+            quote! { None }
+        } else {
+            quote! { Some(#methods) }
+        };
+
+        quote! {
             docs: ::godot::docs::InherentImplDocs {
                 methods: #methods,
-                signals: #signals,
-                constants: #constants,
-            }.into()
-        };
-        Some(field_definition)
+                signals_block: #signals_block,
+                constants_block: #constants_block,
+            }
+        }
     }
-    pieces(functions, signals, constants).unwrap_or_else(|| quote! { docs: None })
+    pieces(functions, signals, constants)
 }
 
 pub fn make_virtual_impl_docs(vmethods: &[ImplMember]) -> TokenStream {
-    match vmethods
+    let virtual_methods = vmethods
         .iter()
         .filter_map(|x| match x {
             venial::ImplMember::AssocFunction(f) => Some(f.clone()),
             _ => None,
         })
-        .map(make_virtual_method_docs)
-        .collect::<Option<String>>()
-    {
-        Some(vmethods) => quote! {
-            virtual_method_docs: #vmethods,
-        },
-        None => quote! {
-            virtual_method_docs: ""
-        },
+        .filter_map(make_virtual_method_docs)
+        .collect::<String>();
+
+    if virtual_methods.is_empty() {
+        quote! { virtual_method_docs: None, }
+    } else {
+        quote! { virtual_method_docs: Some(#virtual_methods), }
     }
```

Comment:
the way this is used seems a bit silly, as its just unwrap_or_default'd later:
```rust
            let methods_block = if methods.is_some() || virtual_methods.is_some() {
                format!(
                    "<methods>{m}{vm}</methods>",
                    m = methods.unwrap_or_default(),
                    vm = virtual_methods.unwrap_or_default())
            } else {
                String::new()
            };
```
can that be changed to just
```rust
(virtual_methods.len() + methods.len() > 0)
   .then(|| format!("<methods>{methods}{virtual_methods}</methods>"))
   .unwrap_or_default()
```
and do this here: (and in the make method docs function)
```suggestion
    quote! { virtual_method_docs: #virtual_methods }
```
?

Reply:
thanks a lot! It actually works fine and looks like the cleanest and the least messy solution